### PR TITLE
feat: allow to have a custom configuration per branches

### PR DESCRIPTION
**Pull request by KillianHmyd**
This could allow to have different behaviours according to the branch from which the release is triggered. For example the following example would allow to have changelog only from `lts/*` and `master` branches in two different channels : 

```js
[
      "semantic-release-slack-bot",
      {
       "notifyOnSuccess": false,
        "notifyOnFail": false,
        "branchesConfig": [
           {
             "branchPattern": "master",
             "notifyOnSuccess": true,
             "notifyOnFail": true,
             "slackWebhook": "stable-channel",
          },
           {
             "branchPattern": "lts/x",
             "notifyOnSuccess": true,
             "notifyOnFail": true,
             "slackWebhook": "lts-channel",
          }
       ]
      }
    ]
```

### CODE REVIEW

**Comment by juliuscc** [comment]
I have a question. I saw you used micromatch. Is that the same pattern-strategy that is used by semantic-release ([see here](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/workflow-configuration.md))? It should work identically I think, as a subtly different behavior would be a large cause of friction for users.

**Comment by KillianHmyd** [comment]
Yes it is exactly the same 👍  If you check the link they put on `glob` it redirects to micromatch documentation.
It is also referenced in [their package.json](https://github.com/semantic-release/semantic-release/blob/master/package.json#L42) and seems to be used to filter the branches [here](https://github.com/semantic-release/semantic-release/blob/e35e5bb73f3628906553709ac0f5beb8f565c61a/lib/branches/expand.js#L11).

File: README.md [hunk 1]
```diff
@@ -45,8 +57,10 @@ The plugin can be configured in the [**semantic-release** configuration file](ht
 
 With this example:
 
-- Slack notifications are skipped on a successful release
-- Slack notifications are sent on a failed release
+- Slack notification will always be sent using the "https://my-webhook.com" webhook url
+- Slack notifications are sent on a failure release from branches matching "lts/\*"
+- Slack notifications are sent on a failure or successful release from branch "master"
+- Slack notifications are skipped on every other branches
```
Comment (juliuscc) [inline README.md:63]: Nitpick:
```suggestion
- Slack notifications are skipped on all other branches
```

File: lib/verifyConditions.js [hunk 2]
```diff
@@ -30,4 +30,28 @@ module.exports = (pluginConfig, context) => {
...
+
+  if (
+    pluginConfig.branchesConfig &&
+    !Array.isArray(pluginConfig.branchesConfig)
+  ) {
+    logger.log('branchesConfig is defined and is not an array')
+    throw new SemanticReleaseError(
+      'branchesConfig is not an array.',
+      'EINVALIDBRANCHCONFIG',
+      `Provided branches configuration is not an array. Make "branchesConfig" is properly set in your configuration option.`
```
Comment (juliuscc) [inline lib/verifyConditions.js:42]: Nitpick:
```suggestion
      `Provided branches configuration is not an array. Ensure "branchesConfig" is properly set in your configuration option.`
```

File: lib/verifyConditions.js [hunk 3]
```diff
@@ -30,4 +30,28 @@ module.exports = (pluginConfig, context) => {
...
+
+  if (
+    pluginConfig.branchesConfig &&
+    pluginConfig.branchesConfig.some(({ pattern }) => !pattern)
+  ) {
+    logger.log('pattern is not defined in branchesConfig')
+    throw new SemanticReleaseError(
+      'pattern is not defined in branchesConfig.',
+      'ENOPATTERN',
+      `A pattern for the branch configuration must be added. Make "branchesConfig" is properly set in your configuration option.`
```
Comment (juliuscc) [inline lib/verifyConditions.js:54]: Nitpick:
```suggestion
      `A pattern for the branch configuration must be added. Ensure "branchesConfig" is properly set in your configuration option.`
```

File: test/getConfigToUse.test.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      notifyOnSuccess: false,
+      notifyOnFail: true
+    }
+    const expectedConfig = pluginConfig
+
+    const actual = getConfigToUse(pluginConfig, context)
+    assert.deepEqual(actual, expectedConfig)
+  })
+
+  it('should return the branche config merged with the global config when branchesConfig match the branch', () => {
```
Comment (juliuscc) [inline test/getConfigToUse.test.js:18]:
```suggestion
  it('should return the branch config merged with the global config when branchesConfig match the branch', () => {
```

**Comment by juliuscc** [comment]
I've looked through the PR and it's great! You have great tests and I think the implementation is clear. It's also great that you expanded the example in the readme.

If you fix the small spelling/grammatical nitpicks I will merge 👍

**Comment by KillianHmyd** [comment]
I've pushed your suggestions, thanks !
Also we tested it in one of our private repository and it worked well 👍

**Review by juliuscc — APPROVED** [review]
Looks great! Thanks! 👍

**Pull request merged**